cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+// healthCheck is a helper struct for easily performing healthchecks against the cluster and printing the output
+type healthCheck struct {
+	description, okMessage, failMessage string
+	// f is invoked with a k8s client passed to it. Should return an optional warning and/or an error
+	f func(clientset.Interface) error
+}
+
+// VerifyClusterHealth makes sure:
+// - the API /healthz endpoint is healthy
+// - all Nodes are Ready
+// - (if self-hosted) that there are DaemonSets with at least one Pod for all control plane components
+// - (if static pod-hosted) that all required Static Pod manifests exist on disk
+func VerifyClusterHealth(client clientset.Interface) bool {
+	fmt.Println("[upgrade] Making sure the cluster is healthy:")
+
+	defaultHealthChecks := []healthCheck{
+		{
+			description: "API Server health",
+			okMessage:   "Healthy",
+			failMessage: "Unhealthy",
+			f:           apiServerHealthy,
+		},
+		{
+			description: "Node health",
+			okMessage:   "All Nodes are healthy",
+			failMessage: "More than one Node unhealthy",
+			f:           nodesHealthy,
+		},
+		// TODO: Add a check for ComponentStatuses here?
+	}
+
+	success := runHealthChecks(client, defaultHealthChecks)
+	if !success {
+		return false
+	}
+
+	// If the control plane is self-hosted, we should run more health checks
+	fmt.Printf("[upgrade/health] Checking if control plane is Static Pod-hosted or Self-Hosted: ")
```

Comment:
Outcommented for now; can revisit later. I think it makes sense to leave this detail out